packages/@aws-cdk/cdk/lib/core/tag-manager.ts

```diff
@@ -87,51 +99,44 @@ export class TagManager {
       // tag is blocked by a remove
       return;
     }
-    delete this.tags[key];
+    delete this.tagSet[key];
     this.removedTags[key] = priority;
   }
 
   /**
    * Renders tags into the proper format based on TagType
    */
-  public renderTags(): any {
-    const keys = Object.keys(this.tags);
+  public renderTags(propertyTags?: any): any {
+    const formatter = this.tagFormatter();
+    const tags: ITag[] = [];
+    for (const key of Object.keys(this.tagSet)) {
+      tags.push(this.tags[this.tagSet[key]]);
+    }
+    return formatter.renderTags(tags, propertyTags);
+  }
+
+  private tagFormatter(): ITagFormatter {
     switch (this.tagType) {
```

Comment:
I preferred to leave the codegen with just the simple enum and moved to the `const TAG_FORMATTERS ... ` pattern. If you strongly otherwise I'll move it to codegen.